extensions/amp-social-share/0.1/amp-social-share.js

```diff
@@ -0,0 +1,138 @@
+ */
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {getSocialConfig} from './amp-social-share-config';
+import {addParamsToUrl} from '../../../src/url.js';
+import {elementByTag} from '../../../src/dom.js';
+
+/** @const {number} */
+const DEFAULT_WIDTH = 60;
+/** @const {number} */
+const DEFAULT_HEIGHT = 44;
+
+/** @const {string} */
+const CONTAINER_TAG = 'span';
+/** @const {string} */
+const LINK_TAG = 'a';
+/** @const {string} */
+const JSON_TAG = 'script';
+
+class AmpSocialShare extends AMP.BaseElement {
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  buildCallback() {
+
+    /** @private @const {Object} */
+    this.config_ = this.getElementConfig_(this.element);
+    AMP.assert(this.config_, 'The <script> config is required for ' +
+      '<amp-social-share> %s');
+
+    /** @private @const {!Element} */
+    this.type_ = AMP.assert(this.element.getAttribute('type'),
+        'The type attribute is required for <amp-social-share> %s',
+        this.element);
+    /** @private @const {!Object} */
+    this.typeConfig_ = getSocialConfig(this.type_);
+
+    /** @private @const {Element} */
+    this.container_ = this.renderShare_(this.config_);
+
+    /** @private @const {number} */
+    this.width = this.element.getAttribute('width') || DEFAULT_WIDTH;
+    /** @private @const {number} */
+    this.height = this.element.getAttribute('height') || DEFAULT_HEIGHT;
+
+    this.element.appendChild(this.container_);
+  }
+
+  /** @override */
+  layoutCallback() {
+    return Promise.resolve();
+  }
+
+  /**
+   * Renders the share based on the provided config.
+   * @param {!Object} config
+   * @return {!Element}
+   */
+  renderShare_(config) {
+    const urlParams = {};
+
+    for (const param in this.typeConfig_['params']) {
+      const paramConf = this.typeConfig_['params'][param];
+      AMP.assert(!paramConf['required'] || param in config,
+        param + ' is a required configuration item for ' + this.type_);
+
+      if (param in config) {
+        const value = config[param];
+        if ('maxlength' in paramConf) {
+          const maxlength = paramConf['maxlength'];
+          AMP.assert(value.length < maxlength,
+            param + ' cannot exceed ' + maxlength);
+        }
+        urlParams[paramConf['param']] = value;
+      } else if (paramConf['type'] == 'fixed') {
+        urlParams[paramConf['param']] = paramConf['value'];
+      }
+    }
+    let container = elementByTag(this.element, CONTAINER_TAG);
+    let link = elementByTag(this.element, LINK_TAG);
+    if (container === null) {
+      container = this.getWin().document.createElement(CONTAINER_TAG);
+      container.classList.add('-amp-social-share');
+      container.classList.add('-amp-social-share-' + this.type_);
+    }
+    if (link === null) {
+      link = this.getWin().document.createElement(LINK_TAG);
+      link.innerHTML = this.typeConfig_['text'];
```

Comment:
Yes, this should definitely be `textContent` here. Few options:
 1. You can research `white-space` CSS rules and how to modify them for your desired needs.
 2. You can set `a` to `display: block` and provide it with width/height to fill the available space. In general this is a good idea since what you do is a block. However, that'd mean that you can't then host `a` inside of `span`. Although, I don't yet understand what `span` does and why it's needed.
 3. You can use JS character equivalent to `&nbsp;`. I want to say it's `\u00A0`, but not 100%.

I'd like to understand the markup and its intent a little better before advising the best approach. But I do lean toward `display:block` on `a` and possible no `span` at all.